### PR TITLE
Unified ast 2

### PR DESCRIPTION
**Pull request by ndemeyembosco**
Code for the generic AST up to where we had it working with pattern synonyms defined for the undecorated AST

### CODE REVIEW

**Comment by SanjitKal** [comment]
I think we should complete the work on this branch (add the extension descriptors for the Typed AST) before we merge this branch into master. I will work on that today.

**Comment by ndemeyembosco** [comment]
Cool

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by ndemeyembosco** [comment]
Hey! have you done this already?? and if so could you push the changes? Thanks 😊

________________________________

File: stack.yaml [hunk 1]
```diff
@@ -10,6 +10,7 @@ packages:
 
 # Packages to be pulled from upstream that are not in the resolver (e.g., acme-missiles-0.3)
 extra-deps:
+- parser-combinators-0.4.0
```
Comment (byorgey) [inline stack.yaml]: Just curious, why is this necessary?

File: src/Disco/Typecheck.hs [hunk 2]
```diff
@@ -919,44 +920,56 @@ inferQual (QGuard (unembed -> t))   = do
...
+  rest <- checkTuplePat ps ty2
+  return ((ctx, apt):rest)
 checkTuplePat ps ty = throwError $ NotTuplePattern (PTup ps) ty
 
 -- | Successfully return the empty context.  A convenience method for
 --   checking patterns that bind no variables.
-ok :: TCM TyCtx
-ok = return emptyCtx
+-- ok :: TCM TyCtx
+-- ok = return emptyCtx
```
Comment (byorgey) [inline src/Disco/Typecheck.hs]: Remember to delete commented-out code like this before we merge.

File: src/Disco/AST/Generic.hs [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  Disco.AST.Surface
+-- Copyright   :  (c) 2016 disco team (see LICENSE)
+-- License     :  BSD-style (see LICENSE)
+-- Maintainer  :  byorgey@gmail.com
+--
+-- Abstract syntax trees representing the surface syntax of the Disco
+-- language.
+--
```
Comment (byorgey) [inline src/Disco/AST/Generic.hs]: This comment still says `Disco.AST.Surface` and talks about surface syntax.

File: src/Disco/AST/Typed.hs [hunk 4]
```diff
@@ -21,132 +28,347 @@
...
 import           Disco.AST.Surface
 import           Disco.Syntax.Operators
 import           Disco.Types
+import           Disco.AST.Generic
 
 -- TODO: Should probably really do this with a 2-level/open recursion
 -- approach, with a cofree comonad or whatever
 
 -- | An @ATerm@ is a typechecked term where every node in the tree has
 --   been annotated with the type of the subterm rooted at that node.
```
Comment (byorgey) [inline src/Disco/AST/Typed.hs]: Move this comment to be right above the definition of `ATerm`, and add a comment about `TY` being the descriptor

File: src/Disco/AST/Typed.hs [hunk 5]
```diff
@@ -21,132 +28,347 @@
...
+pattern ATList :: Type -> [ATerm] -> Maybe (Ellipsis ATerm) -> ATerm
+pattern ATList ty termlist mellipses = TList_ ty termlist mellipses
+
+pattern ATListComp :: Type -> Bind (Telescope AQual) ATerm -> ATerm
+pattern ATListComp ty bind = TListComp_ ty bind
+
+pattern ATAscr :: ATerm -> Type -> ATerm
+pattern ATAscr term ty = TAscr_ () term ty
+
+-- data ATerm where
```
Comment (byorgey) [inline src/Disco/AST/Typed.hs]: Be sure to delete all this commented-out code.

File: stack.yaml (hunk 1 above)
Comment (SanjitKal) [reply stack.yaml]: This comes from the commit titled "Integrate Generic.hs with Surface.hs". I am not sure why it is needed, but if I remove it, the project fails to build with the following error message:

"--  While building custom Setup.hs for package disco-0.1.0.0 using:
      /Users/sanjitkalapatapu1/.stack/setup-exe-cache/x86_64-osx/Cabal-simple_mPHDZzAJ_2.0.1.0_ghc-8.2.2 --builddir=.stack-work/dist/x86_64-osx/Cabal-2.0.1.0 build lib:disco exe:disco test:disco-examples test:disco-tests --ghc-options " -ddump-hi -ddump-to-file -fdiagnostics-color=always"
    Process exited with code: ExitFailure 1"

@ndemeyembosco , do you know why this was added?

Comment (byorgey) [reply stack.yaml]: It I comment that line out it still builds fine for me.  What version of `stack` do you have?

Comment (SanjitKal) [reply stack.yaml]: Version 1.7.1

**Comment by byorgey** [comment]
What's the status of this?  Do you think it is ready to merge?  Note there is a bit of extra work to be done now, since we have merged in some of Jacob's work on adding sets into `master`.  That modified the AST so we will have to make the changes generic.  Probably the best approach is to merge `master` into this branch first, fix the merge conflicts, make sure everything builds and the tests still pass, etc., and then merge back into `master` once that's done.

File: src/Disco/AST/Generic.hs [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+  -- | Pattern guard (@when term = pat@)
+  {-I am thinking about changing Pattern_ e here to Pattern_ b 
+    but you wouldn't be able to derive Generic if you do that. -}
+  GPat_  :: X_GPat e -> Embed (Term_ e) -> Pattern_ e -> Guard_ e
+
+  deriving Generic
+
+deriving instance Forall_t Show  e => Show (Guard_ e)
+
+-- | Patterns.
```
Comment (byorgey) [inline src/Disco/AST/Generic.hs]: This comment should move to right above `data Pattern_`.  At the moment it is attached to `type family X_PVar`.

File: src/Disco/AST/Generic.hs [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+type family X_GPat e
+
+data Guard_ e where
+
+  -- | Boolean guard (@if <test>@)
+  GBool_ :: X_GBool e -> Embed (Term_ e) -> Guard_ e
+
+  -- | Pattern guard (@when term = pat@)
+  {-I am thinking about changing Pattern_ e here to Pattern_ b 
+    but you wouldn't be able to derive Generic if you do that. -}
```
Comment (byorgey) [inline src/Disco/AST/Generic.hs]: Can we get rid of this comment?

File: src/Disco/AST/Generic.hs [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+-- | An ellipsis is an "omitted" part of a literal list, of the form
+--   @..@ or @.. t@.
+data Ellipsis t where
+  Forever ::      Ellipsis t   -- @..@
+  Until   :: t -> Ellipsis t   -- @.. t@
+  deriving (Show, Generic, Functor, Foldable, Traversable)
+
+instance Alpha t => Alpha (Ellipsis t)
+
+-- Note: very similar to guards-
```
Comment (byorgey) [inline src/Disco/AST/Generic.hs]: We should probably delete this comment.

File: src/Disco/AST/Generic.hs [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+
+-- | A container is a wrapper for sets and lists.
+data Container where
+  CList :: Container
+  CSet :: Container
+  deriving (Show, Eq, Enum, Generic)
+
+instance Alpha Container
+
+-- | An ellipsis is an "omitted" part of a literal list, of the form
```
Comment (byorgey) [inline src/Disco/AST/Generic.hs]: Instead of "literal list" this should now say something like "literal container (such as a list or set)".

File: src/Disco/AST/Generic.hs [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+  TContainerComp_ :: X_TContainerComp e -> Container -> Bind (Telescope (Qual_ e)) (Term_ e) -> Term_ e
+
+  -- | Type ascription, @(Term_ e : type)@.
+  TAscr_  :: X_TAscr e -> Term_ e -> Type -> Term_ e
+
+  XTerm_   :: X_Term e -> Term_ e 
+  deriving (Generic)
+
+ 
+type Forall_t (a :: * -> Constraint) e
```
Comment (byorgey) [inline src/Disco/AST/Generic.hs]: Can we add a Haddock comment (i.e. a comment beginning with `-- |`) right before `Forall_t` explaining what it is for?

File: src/Disco/AST/Generic.hs [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+  -- | A containter for sets and lsits.
+  TContainer_ :: X_TContainer e -> Container -> [Term_ e] -> Maybe (Ellipsis (Term_ e)) -> Term_ e
+
+  -- | A container for set and list comprehensions
+  TContainerComp_ :: X_TContainerComp e -> Container -> Bind (Telescope (Qual_ e)) (Term_ e) -> Term_ e
+
+  -- | Type ascription, @(Term_ e : type)@.
+  TAscr_  :: X_TAscr e -> Term_ e -> Type -> Term_ e
+
+  XTerm_   :: X_Term e -> Term_ e
```
Comment (byorgey) [inline src/Disco/AST/Generic.hs]: Needs a Haddock comment.

File: src/Disco/AST/Generic.hs [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+      , Qual_ (..)
+      , X_QBind
+      , X_QGuard
+      , Binding_ (..)
+      , Branch_
+      , Guard_ (..)
+      , X_GBool
+      , X_GPat
+      , Forall_t
+      , Telescope (..)
```
Comment (byorgey) [inline src/Disco/AST/Generic.hs]: Let's move `Telescope` in the export list to be right next to the related functions.

**Comment by byorgey** [comment]
Why did `reference/type-inference-with-subtypes.pdf` get deleted?  Please restore it.

File: src/Disco/AST/Generic.hs [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TemplateHaskell       #-}
+{-# LANGUAGE UndecidableInstances  #-}
+{-# LANGUAGE ViewPatterns          #-}
+{-# LANGUAGE TypeOperators, PatternSynonyms #-}
+{-# LANGUAGE EmptyCase, StandaloneDeriving #-}
+{-# LANGUAGE TypeFamilies, DataKinds, ConstraintKinds #-}
```
Comment (byorgey) [inline src/Disco/AST/Generic.hs]: Let's clean this up (align all the `#-}`, one extension per line, and alphabetize).  You may want to try the [`stylish-haskell` tool](http://hackage.haskell.org/package/stylish-haskell).

File: src/Disco/AST/Surface.hs [hunk 14]
```diff
@@ -9,7 +9,9 @@
 {-# LANGUAGE TemplateHaskell       #-}
 {-# LANGUAGE UndecidableInstances  #-}
 {-# LANGUAGE ViewPatterns          #-}
-{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE TypeOperators, PatternSynonyms #-}
+{-# LANGUAGE EmptyCase, StandaloneDeriving #-}
+{-# LANGUAGE TypeFamilies, DataKinds, ConstraintKinds #-}
```
Comment (byorgey) [inline src/Disco/AST/Surface.hs]: Style cleanup

File: src/Disco/Typecheck.hs [hunk 15]
```diff
@@ -1090,13 +1101,25 @@ erase (ATAscr at ty)        = TAscr (erase at) ty
...
+erasePattern (APVar n) = PVar (coerce n)
+erasePattern APWild = PWild
+erasePattern APUnit = PUnit
+erasePattern (APBool b) = PBool b
+erasePattern (APTup alp) = PTup $ map erasePattern alp
+erasePattern (APInj s apt) = PInj s (erasePattern apt)
+erasePattern (APNat n) = PNat n
+erasePattern (APSucc apt) = PSucc $ erasePattern apt
+erasePattern (APCons ap1 ap2) = PCons (erasePattern ap1) (erasePattern ap2)
+erasePattern (APList alp) = PList $ map erasePattern alp
```
Comment (byorgey) [inline src/Disco/Typecheck.hs]: Can we line up all these `=` ?

File: src/Disco/AST/Generic.hs (hunk 13 above)
Comment (SanjitKal) [reply src/Disco/AST/Generic.hs]: Thanks for showing stylish-haskell. It's great.

Comment (byorgey) [reply src/Disco/AST/Generic.hs]: Yeah, I actually have my editor set up to automatically run stylish-haskell every time I save!
